### PR TITLE
EnColMan OK

### PR DESCRIPTION
**Pull request by kyleburnette**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

This actor is a mess, so I didn't do all that much documentation. It's the lab HP spawned by the fish, the garo master falling rocks, and the garo master bomb that spawns at the end of the fight for some reason. Truly remarkable.

### CODE REVIEW

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 1]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
     (ActorFunc)NULL,
 };
 
-#endif
+static Color_RGBA8 primColor = { 0x3C, 0x32, 0x14, 0xFF };
+static Color_RGBA8 envColor = { 0x28, 0x1E, 0x1E, 0xFF };
+
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:66]: Colours in dec, please.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 2]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
...
+}
+
+void EnColMan_Destroy(Actor* thisx, GlobalContext* globalCtx) {
+    EnColMan* this = THIS;
+
+    Collider_DestroyCylinder(globalCtx, &this->collider);
+}
+
+void func_80AFDD60(EnColMan* this) {
+    if (!(gSaveContext.weekEventReg[0x38] & 2)) {
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:103]: I *think* we do dec for these: it is an array access, after all.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 3]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
...
+        }
+        this->unk148 = 1;
+    }
+    this->actionFunc = func_80AFDE00;
+}
+
+void func_80AFDE00(EnColMan* this, GlobalContext* globalCtx) {
+    if (this->actor.bgCheckFlags & 1) {
+        if (this->actor.params == EN_COL_MAN_HEART_PIECE) {
+            this->actor.params = EN_COL_MAN_RECOVERY_HEART;
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:125]: This is weird enough that I wonder if something else is going on here.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 4]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
...
+            velocity.x = randPlusMinusPoint5Scaled(2.0f);
+            velocity.y = Rand_ZeroFloat(2.0f) + 1.0f;
+            velocity.z = randPlusMinusPoint5Scaled(2.0f);
+            accel.y = -0.1f;
+            accel.z = 0.0f;
+            accel.x = 0.0f;
 
-extern ColliderCylinderInit D_80AFE730;
+            func_800B0EB0(globalCtx, &this->actor.world.pos, &velocity, &accel, &primColor, &envColor,
+                          Rand_ZeroFloat(50.0f) + 60.0f, 0x1E, Rand_ZeroFloat(5.0f) + 20.0f);
```
Comment (EllipticEllipsis) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:189]: The 0x1E is almost certainly actually 30, although it's not clear precisely what this function does. Timer or params for something would be my guess.

**Review by EllipticEllipsis — COMMENTED** [review]
Nice job, just a few style things. But while I can kind of see why you wouldn't want to try and document this, you should be able to look at the displaylists in Z64Utils, even if they are in gameplay_keep. From that you can work out what draw function is what, and probably get a few more of the functions named, at least.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 5]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
...
+    this->scale = BREG(55) * 0.01f + 0.05f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/func_80AFDE00.s")
+    if (BREG(60) || (this->actor.world.rot.z != 0)) {
+        if (this->actor.params == EN_COL_MAN_BOMB) {
+            Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_CLEAR_TAG, this->actor.parent->world.pos.x,
+                        this->actor.parent->world.pos.y, this->actor.parent->world.pos.z, 0, 0, 0, 0);
+        } else {
+            EnBom* bomb = (EnBom*)Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_BOM, this->actor.world.pos.x,
+                                              this->actor.world.pos.y, this->actor.world.pos.z, 0, 0, 0, 0);
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:211]: This got me interested so I did some digging in KZ. It looks like what you have as `EN_COL_MAN_BOMB` is the "bomb" used by the Garo Master in stone tower (note it spawns an 'ACTOR_EN_CLEAR_TAG` so actually doesn't affect Link). `EN_COL_MAN_UNK4` are the bombs used by the other Garo Masters (On the moon and in secret shrine). These are real bombs that can hurt Link.

Comment (kyleburnette) [reply src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:211]: Ah nice, I knew something was weird about it. I spawned in both using kz and didn't see much of a difference and didn't look far enough into it

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c (hunk 3 above)
Comment (kyleburnette) [reply src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:125]: I thought this was weird as well, not gonna lie

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 6]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/func_80AFDD60.s")
+void func_80AFE25C(EnColMan* this, GlobalContext* globalCtx) {
+    this->scale = BREG(55) * 0.01f + 0.05f;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/func_80AFDE00.s")
+    if (BREG(60) || (this->actor.world.rot.z != 0)) {
+        if (this->actor.params == EN_COL_MAN_CUTSCENE_BOMB) {
+            Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_CLEAR_TAG, this->actor.parent->world.pos.x,
+                        this->actor.parent->world.pos.y, this->actor.parent->world.pos.z, 0, 0, 0, 0);
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:208]:
```suggestion
            Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_CLEAR_TAG, this->actor.parent->world.pos.x,
                        this->actor.parent->world.pos.y, this->actor.parent->world.pos.z, 0, 0, 0, CLEAR_TAG_SMALL_EXPLOSION);
```

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 7]
```diff
@@ -39,34 +61,220 @@ const ActorInit En_Col_Man_InitVars = {
...
+        }
+        this->unk148 = 0;
+    } else {
+        this->actor.draw = func_80AFE4AC;
+        this->actor.shape.yOffset = 300.0f;
+        this->actor.shape.shadowScale = 5.0f;
+        if (this->actor.params == EN_COL_MAN_HEART_PIECE) {
+            this->actor.gravity = -2.0f;
+        }
+        this->unk148 = 1;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:117]:
```suggestion
        this->unk148 = EN_COL_MAN_HEART_PIECE;
    } else {
        this->actor.draw = func_80AFE4AC;
        this->actor.shape.yOffset = 300.0f;
        this->actor.shape.shadowScale = 5.0f;
        if (this->actor.params == EN_COL_MAN_HEART_PIECE) {
            this->actor.gravity = -2.0f;
        }
        this->unk148 = EN_COL_MAN_RECOVERY_HEART;
```
Taking a read through the code, it looks like `this->unk148` stores the type of actor this represents and aligns with the enum you've made below. Thoughts?

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.h [hunk 8]
```diff
@@ -8,10 +8,22 @@ struct EnColMan;
...
 typedef struct EnColMan {
-    /* 0x0000 */ Actor actor;
-    /* 0x0144 */ EnColManActionFunc actionFunc;
-    /* 0x0148 */ char unk_144[0x58];
-} EnColMan; // size = 0x1A0
+    /* 0x000 */ Actor actor;
+    /* 0x144 */ EnColManActionFunc actionFunc;
+    /* 0x148 */ s16 unk148; // never used, only set
+    /* 0x14A */ char unk14A[0x2];
+    /* 0x14C */ s32 hasSetRandomValues;
```
Comment (engineer124) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.h:15]:
```suggestion
    /* 0x148 */ s16 unk148; // never used, only set
    /* 0x14C */ s32 hasSetRandomValues;
```
Could this just be padding?

**Review by engineer124 — COMMENTED** [review]
Nice work! some small comments

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 9]
```diff
@@ -39,34 +61,221 @@ const ActorInit En_Col_Man_InitVars = {
...
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/func_80AFE25C.s")
+void func_80AFE414(Actor* thisx, GlobalContext* globalCtx) {
+    EnColMan* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/EnColMan_Update.s")
+    OPEN_DISPS(globalCtx->state.gfxCtx);
+    func_8012C2DC(globalCtx->state.gfxCtx);
+    func_800B8118(&this->actor, globalCtx, 0);
+    gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    gSPDisplayList(POLY_XLU_DISP++, D_0405AAB0);
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:243]: From gameplay_keep, should include its object header and use the proper symbol.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 10]
```diff
@@ -39,34 +61,221 @@ const ActorInit En_Col_Man_InitVars = {
...
+void func_80AFE4AC(Actor* thisx, GlobalContext* globalCtx) {
+    EnColMan* this = THIS;
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/func_80AFE4AC.s")
+    func_8012C2DC(globalCtx->state.gfxCtx);
+    func_8012C28C(globalCtx->state.gfxCtx);
+    OPEN_DISPS(globalCtx->state.gfxCtx);
+    POLY_OPA_DISP = func_801660B8(globalCtx, POLY_OPA_DISP);
+    POLY_OPA_DISP = func_8012C724(POLY_OPA_DISP);
+    gSPSegment(POLY_OPA_DISP++, 0x08, Lib_SegmentedToVirtual(&D_0405E6F0));
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:255]: From gameplay_keep, should include its object header and use the proper symbol.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 11]
```diff
@@ -39,34 +61,221 @@ const ActorInit En_Col_Man_InitVars = {
...
+    gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    gSPDisplayList(POLY_OPA_DISP++, D_0405F6F0);
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:257]: From gameplay_keep, should include its object header and use the proper symbol.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 12]
```diff
@@ -39,34 +61,221 @@ const ActorInit En_Col_Man_InitVars = {
...
 
-#pragma GLOBAL_ASM("asm/non_matchings/overlays/ovl_En_Col_Man/func_80AFE650.s")
+void func_80AFE650(Actor* thisx, GlobalContext* globalCtx) {
+    func_8012C28C(globalCtx->state.gfxCtx);
+    OPEN_DISPS(globalCtx->state.gfxCtx);
+    POLY_OPA_DISP = func_801660B8(globalCtx, POLY_OPA_DISP);
+    POLY_OPA_DISP = func_8012C724(POLY_OPA_DISP);
+    gSPSegment(POLY_OPA_DISP++, 0x08, Lib_SegmentedToVirtual(&D_0405CEF0));
+    gSPMatrix(POLY_OPA_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx), G_MTX_NOPUSH | G_MTX_LOAD | G_MTX_MODELVIEW);
+    gSPDisplayList(POLY_OPA_DISP++, D_0405F6F0);
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:279]: From gameplay_keep, so should include its object header and use the proper symbol.

File: src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c [hunk 13]
```diff
@@ -39,34 +61,221 @@ const ActorInit En_Col_Man_InitVars = {
     (ActorFunc)NULL,
 };
 
-#endif
+static Color_RGBA8 primColor = { 60, 50, 20, 255 };
+static Color_RGBA8 envColor = { 40, 30, 30, 255 };
+
+extern Gfx D_0405AAB0[]; // heart piece
+extern Gfx D_0405F6F0[];
+extern Gfx D_0406AB30[];
```
Comment (hensldm) [inline src/overlays/actors/ovl_En_Col_Man/z_en_col_man.c:69]: Wouldn't need to extern these when including gameplay keep.

**Review by hensldm — COMMENTED** [review]
Just some gameplay keep symbols left to fix.

**Review by hensldm — APPROVED** [review]

**Review by AngheloAlf — APPROVED** [review]

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**